bochs/iodev/usb/scsi_device.cc

```diff
@@ -789,6 +797,166 @@ Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, int lun, bool as
+              p[2] = 0;  // Start slot
+              p[3] = 0;  // End slot
+              out_len += 4;  // 8 total bytes
+            } else
+            if (buf[4] & (1<<5)) { // Multi-Initiator request bit is set
+              Class = 5;
+              //p[0] = 0; // Bits 3:0 = Event Code (must be zero until an event occurs, then zero after read until next event occurs)
+            } else
+            if (buf[4] & (1<<6)) { // Device Busy request bit is set
+              Class = 6;
+              //p[0] = 0; // Bits 3:0 = Event Code (must be zero until an event occurs, then zero after read until next event occurs)
+            }
+            if (buf[4] & (1<<7)) { // Reserved bit is set (MMC4R05A.PDF, 6.7.1.3, page 242, not considered an error)
+              BX_DEBUG(("Event Status Notify: bit 7 in the request byte is set"));
+              Class = 7;
+            }
+          }
+          // only send the allocated length requested
+          if (out_len > len)
+            out_len = len;
+          
+          // out buffer
+          // Header is four bytes in length
+          r->buf_len = (int) out_len;
+          out_len = (out_len >= 4) ? (out_len - 4) : 0; // count of bytes to follow the header
+          outbuf[0] = ((out_len >> 8) & 0xFF); // this is the out_len (the count of bytes sent to the guest) (big-endian)
+          outbuf[1] = ((out_len >> 0) & 0xFF);
+          outbuf[2] = 0x00 | Class; // bit 7 = 0, bits 6:3 = resv, bits 2:0 = class returned
+          outbuf[3] = 0b00010000;  // supported Event Classes (only event class 4 supported at this time)
+        } else { // polled bit
+          // we don't (currently) support asynchronous operation.
+          goto fail;
+        }
+      }
+      break;
+    case 0x51:  // SCSI_CD_READ_DISC_INFO
+      {
+        Bit16u out_len;
+        
+        // This code is experimental. may need some help
+        BX_DEBUG(("Read Disc Information: Data type requested 0x%02X", buf[1] & 7));
+        
+        out_len = (Bit32s) (((Bit32u) buf[7] << 8) | buf[8]);
+        switch (buf[1] & 7) {
+          // mmc6r02f.pdf, section 6.21.4, page 379(427) ??
+          case 0: // Standard Disk Information
+            outbuf[ 0] = (32 * (0 * 8)) >> 8; // msb length 32 + (8 * number of OPC tables)
+            outbuf[ 1] = (32 * (0 * 8)) >> 0; // lsb length
+            outbuf[ 2] = (0 << 5) | (0 << 4) | (3 << 2) | (2 << 0); // Type = 0, Eraseable = 0, State of last session = 3 (Complete), Disc status = 2 (Finalized Disc)
+            outbuf[ 3] = 1; // number of first track on disc (1 based)
+            outbuf[ 4] = 1; // number of sessions (LSB)
+            outbuf[ 5] = 1; // first track number in last session (LSB)
+            outbuf[ 6] = 1; // last track number in last session (LSB)
+            outbuf[ 7] = (0 << 7) | (0 << 6) | (1 << 5) | (0 << 4) | (0 << 3)  | (0 << 2) | (0 << 0); // DID_V, DBC_V, URU, DAC_V, R, Legacy?, BG Format Status
```

Comment:
again, suggest to have defines for all these bits
see examples in CPU code

Reply:
Suggestion taken. Please see previous comments. I will take the time to add them.